### DISCUSSION TITLE
Extra slots for components in general ?

### DISCUSSION THREAD

**Discussion by Namidas** [quasarframework/quasar#17609] [Ideas / Proposals]
The title says it all (I believe), and I've been chewing on this idea for quite some time (although more often than not I've managed to solve this with what I had and/or coding some wrapper compos), but here are 2 very specific (and related) examples that I just couldn't find a solution to:

1) QTable's "records per page" thing...well, actually, this one did have a solution, but I had to rewrite/re-create the whole thing (bottom slot) to be able to change just that.

2) QPagination has no slot whatsoever, and it becomes a problem when using prop `input`, there's no tooltip whatsoever (neither native nor compo) to what that thing is doing, further more, not even a border on the input (ok, you can change that through props and CSS) and it becomes quite obscure...

So...extra slots for components in general, really needed.
Not an expert on Vüe3, and I have no idea how much it costs in terms of processing time, but why not just "abuse" slots?
This is what I usually do (and would have done with QPagination's `<input />`)

```
<slot name="input:pre" />
<slot name="input">
	<div class="input-wrapper">
		<slot name="input:inner:pre" />
		<slot name="input:inner">
			<input />
		</slot>
		<slot name="input:inner:post" />
	</div>
</slot>
<slot name="input:post" />
```

You may or may not use that input wrapper div, but you get the point I guess...

Any thoughts on this ?

Also, last but not least...since Quasar renders components programatically, why not use some sort of "render slots" prop to set up rendering order? Not in all cases, but there are a lot of situations where this would be quite useful and (I think) not that hard to implement, going back to QTable's bottom slot, and assuming we had separate slots for each part of it (named in order of appearance in this example as "rows-per-page", "rows-current", "pagination"), you could have a QTable prop named (for instance) "slotsOrder", an array that defaults to `["rows-per-page","rows-current","pagination"]`, and be able to just feed any other order (or even not include all slot names, for conditional rendering, instead of having to define an empty slot or having an extra property just to void a specific slot)

Just saying..

Anyways, excellent work, totally inlove with the Q ♥

**Comment by stefanvanherwijnen** [comment]
I think the problem with your solution is that it might make the source code a big mess while only solving a few edge cases. Just a single slot with a default definition in the documentation works fine imo.
For QPagination `input` there probably should be an `#input` slot for custom inputs, but I think it can only be replaced with a [proper QInput](https://github.com/quasarframework/quasar/blob/a6b812ee1ce2e0147c6326546204579fa94dc09e/ui/src/components/pagination/QPagination.js#L427) or it won't work at all. I guess a PR solving this problem will be merged, but you should probably ask someone of the Quasar team first.
